File: src/IgnoresAccessChecksToGenerator.VisualBasic.Tasks/PublicizeInternals.vb
Imports System.IO
Imports Microsoft.Build.Framework
Imports Microsoft.Build.Utilities
Imports Mono.Cecil

Public Class PublicizeInternals
	Inherits Task

	Private ReadOnly _resolver As New AssemblyResolver

	<Required>
	Public Property SourceReferences As ITaskItem()

	<Required>
	Public Property AssemblyNames As ITaskItem()

	<Required>
	Public Property IntermediateOutputPath As String

	<Required>
	Public Property GeneratedCodeFilePath() As String

	Public Property ExcludeTypeNames() As ITaskItem()

	Public Overrides Function Execute() As Boolean
		If SourceReferences Is Nothing Then
			Throw New ArgumentNullException(NameOf(SourceReferences))
		End If

		Dim assemblyNameSet As New HashSet(Of String)(AssemblyNames.Select(Function(t) t.ItemSpec), StringComparer.OrdinalIgnoreCase)

		If assemblyNameSet.Count = 0 Then
			Return True
		End If

		Dim excludedTypeNames As HashSet(Of String)
		If ExcludeTypeNames IsNot Nothing Then
			excludedTypeNames = New HashSet(Of String)(ExcludeTypeNames.Select(Function(t) t.ItemSpec), StringComparer.OrdinalIgnoreCase)
		Else
			excludedTypeNames = New HashSet(Of String)(StringComparer.OrdinalIgnoreCase)
		End If

		Dim targetPath = IntermediateOutputPath
		Directory.CreateDirectory(targetPath)

		GenerateAttributes(targetPath, assemblyNameSet)

		For Each assemblyPath In SourceReferences.Select(Function(a) Path.GetDirectoryName(GetFullFilePath(targetPath, a.ItemSpec)))
			_resolver.AddSearchDirectory(assemblyPath)
		Next

		For Each assembly In SourceReferences
			Dim assemblyPath = GetFullFilePath(targetPath, assembly.ItemSpec)
			Dim assemblyName = Path.GetFileNameWithoutExtension(assemblyPath)
			If assemblyNameSet.Contains(assemblyName) Then
				Dim targetAssemblyPath = Path.Combine(targetPath, Path.GetFileName(assemblyPath))

				CreatePublicAssembly(assemblyPath, targetAssemblyPath, excludedTypeNames)
				Log.LogMessageFromText("Created publicized assembly at " & targetAssemblyPath, MessageImportance.Normal)
			End If
		Next assembly

		Return True
	End Function

	Private Sub GenerateAttributes(path As String, asmNames As IEnumerable(Of String))
		Dim attributes = String.Join(Environment.NewLine, asmNames.Select(Function(a) $"<Assembly: System.Runtime.CompilerServices.IgnoresAccessChecksTo(""{a}"")>"))

		Dim content = attributes & "
Namespace Global.System.Runtime.CompilerServices
	<AttributeUsage(AttributeTargets.Assembly, AllowMultiple := True)>
	Friend NotInheritable Class IgnoresAccessChecksToAttribute
		Inherits Attribute

		Sub New(assemblyName As String)
		End Sub
	End Class
End Namespace
"
		File.WriteAllText(GeneratedCodeFilePath, content)

		Log.LogMessageFromText("Generated IgnoresAccessChecksTo attributes", MessageImportance.Low)
	End Sub

	Private Sub CreatePublicAssembly(source As String, target As String, excludedTypeNames As HashSet(Of String))
		Dim assembly = AssemblyDefinition.ReadAssembly(source, New ReaderParameters With {.AssemblyResolver = _resolver})

		For Each [module] In assembly.Modules
			For Each type In [module].GetTypes().Where(Function(type1) Not excludedTypeNames.Contains(type1.FullName))

				If Not type.IsNested AndAlso type.IsNotPublic Then
					type.IsPublic = True
				ElseIf type.IsNestedAssembly OrElse type.IsNestedFamilyOrAssembly OrElse type.IsNestedFamilyAndAssembly Then
					type.IsNestedPublic = True
				End If

				For Each field In type.Fields
					If field.IsAssembly OrElse field.IsFamilyOrAssembly OrElse field.IsFamilyAndAssembly Then
						field.IsPublic = True
					End If
				Next field

				For Each method In type.Methods
					If method.IsAssembly OrElse method.IsFamilyOrAssembly OrElse method.IsFamilyAndAssembly Then
						method.IsPublic = True
					End If
				Next method
			Next type
		Next [module]

		assembly.Write(target)
	End Sub

	Private Function GetFullFilePath(basePath As String, path As String) As String
		Return If(System.IO.Path.IsPathRooted(path), System.IO.Path.GetFullPath(path), System.IO.Path.Combine(basePath, path))
	End Function

	Private Class AssemblyResolver
		Implements IAssemblyResolver

		Private ReadOnly _directories As New HashSet(Of String)(StringComparer.OrdinalIgnoreCase)

		Public Sub AddSearchDirectory(directory As String)
			_directories.Add(directory)
		End Sub

		Public Function Resolve(name As AssemblyNameReference) As AssemblyDefinition Implements IAssemblyResolver.Resolve
			Return Resolve(name, New ReaderParameters)
		End Function

		Public Function Resolve(name As AssemblyNameReference, parameters As ReaderParameters) As AssemblyDefinition Implements IAssemblyResolver.Resolve
			Dim assembly = SearchDirectory(name, _directories, parameters)
			If assembly IsNot Nothing Then
				Return assembly
			End If

			Throw New AssemblyResolutionException(name)
		End Function

		Public Sub Dispose() Implements IDisposable.Dispose
		End Sub

		Private Function SearchDirectory(name As AssemblyNameReference, directories As IEnumerable(Of String), parameters As ReaderParameters) As AssemblyDefinition
			Dim extensions = If(name.IsWindowsRuntime, {".winmd", ".dll"}, {".exe", ".dll"})
			For Each directory In directories
				For Each extension In extensions
					Dim file = Path.Combine(directory, name.Name & extension)
					If Not System.IO.File.Exists(file) Then
						Continue For
					End If
					Try
						Return GetAssembly(file, parameters)
					Catch e1 As BadImageFormatException
					End Try
				Next extension
			Next directory

			Return Nothing
		End Function

		Private Function GetAssembly(file As String, parameters As ReaderParameters) As AssemblyDefinition
			If parameters.AssemblyResolver Is Nothing Then
				parameters.AssemblyResolver = Me
			End If

			Return ModuleDefinition.ReadModule(file, parameters).Assembly
		End Function
	End Class
End Class

File: src/IgnoresAccessChecksToGenerator.Tasks/PublicizeInternals.cs
﻿using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.Build.Framework;
using Microsoft.Build.Utilities;
using Mono.Cecil;

namespace IgnoresAccessChecksToGenerator.Tasks
{
    public class PublicizeInternals : Task
    {
        private readonly AssemblyResolver _resolver = new AssemblyResolver();

        [Required]
        public ITaskItem[] SourceReferences { get; set; }

        [Required]
        public ITaskItem[] AssemblyNames { get; set; }

        [Required]
        public string IntermediateOutputPath { get; set; }

        [Required]
        public string GeneratedCodeFilePath { get; set; }

        public ITaskItem[] ExcludeTypeNames { get; set; }

        public override bool Execute()
        {
            if (SourceReferences == null) throw new ArgumentNullException(nameof(SourceReferences));

            var assemblyNames = new HashSet<string>(AssemblyNames.Select(t => t.ItemSpec), StringComparer.OrdinalIgnoreCase);

            if (assemblyNames.Count == 0)
            {
                return true;
            }

            var excludedTypeNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            if (ExcludeTypeNames != null)
            {
                excludedTypeNames = new HashSet<string>(ExcludeTypeNames.Select(t => t.ItemSpec), StringComparer.OrdinalIgnoreCase);
            }

            var targetPath = IntermediateOutputPath;
            Directory.CreateDirectory(targetPath);

            GenerateAttributes(targetPath, assemblyNames);

            foreach (var assemblyPath in SourceReferences
                .Select(a => Path.GetDirectoryName(GetFullFilePath(targetPath, a.ItemSpec))))
            {
                _resolver.AddSearchDirectory(assemblyPath);
            }

            foreach (var assembly in SourceReferences)
            {
                var assemblyPath = GetFullFilePath(targetPath, assembly.ItemSpec);
                var assemblyName = Path.GetFileNameWithoutExtension(assemblyPath);
                if (assemblyNames.Contains(assemblyName))
                {
                    var targetAssemblyPath = Path.Combine(targetPath, Path.GetFileName(assemblyPath));

                    CreatePublicAssembly(assemblyPath, targetAssemblyPath, excludedTypeNames);
                    Log.LogMessageFromText("Created publicized assembly at " + targetAssemblyPath, MessageImportance.Normal);
                }
            }

            return true;
        }

        private void GenerateAttributes(string path, IEnumerable<string> assemblyNames)
        {
            var attributes = string.Join(Environment.NewLine,
                assemblyNames.Select(a => $@"[assembly: System.Runtime.CompilerServices.IgnoresAccessChecksTo(""{a}"")]"));

            var content = attributes + @"

namespace System.Runtime.CompilerServices
{
    [AttributeUsage(AttributeTargets.Assembly, AllowMultiple = true)]
    internal sealed class IgnoresAccessChecksToAttribute : Attribute
    {
        public IgnoresAccessChecksToAttribute(string assemblyName)
        {
        }
    }
}";
            File.WriteAllText(GeneratedCodeFilePath, content);

            Log.LogMessageFromText("Generated IgnoresAccessChecksTo attributes", MessageImportance.Low);
        }

        private void CreatePublicAssembly(string source, string target, HashSet<string> excludedTypeNames)
        {
            var assembly = AssemblyDefinition.ReadAssembly(source,
                new ReaderParameters { AssemblyResolver = _resolver });

            foreach (var module in assembly.Modules)
            {
                foreach (var type in module.GetTypes().Where(type => !excludedTypeNames.Contains(type.FullName)))
                {
                    if (!type.IsNested && type.IsNotPublic)
                    {
                        type.IsPublic = true;
                    }
                    else if (type.IsNestedAssembly ||
                             type.IsNestedFamilyOrAssembly ||
                             type.IsNestedFamilyAndAssembly)
                    {
                        type.IsNestedPublic = true;
                    }

                    foreach (var field in type.Fields)
                    {
                        if (field.IsAssembly ||
                            field.IsFamilyOrAssembly ||
                            field.IsFamilyAndAssembly)
                        {
                            field.IsPublic = true;
                        }
                    }

                    foreach (var method in type.Methods)
                    {
                        if (method.IsAssembly ||
                            method.IsFamilyOrAssembly ||
                            method.IsFamilyAndAssembly)
                        {
                            method.IsPublic = true;
                        }
                    }
                }
            }

            assembly.Write(target);
        }

        private string GetFullFilePath(string basePath, string path) =>
            Path.IsPathRooted(path) ? Path.GetFullPath(path) : Path.Combine(basePath, path);

        private class AssemblyResolver : IAssemblyResolver
        {
            private readonly HashSet<string> _directories = new HashSet<string>(StringComparer.OrdinalIgnoreCase);

            public void AddSearchDirectory(string directory)
            {
                _directories.Add(directory);
            }

            public AssemblyDefinition Resolve(AssemblyNameReference name)
            {
                return Resolve(name, new ReaderParameters());
            }

            public AssemblyDefinition Resolve(AssemblyNameReference name, ReaderParameters parameters)
            {
                var assembly = SearchDirectory(name, _directories, parameters);
                if (assembly != null)
                {
                    return assembly;
                }

                throw new AssemblyResolutionException(name);
            }

            public void Dispose()
            {
            }

            private AssemblyDefinition SearchDirectory(AssemblyNameReference name, IEnumerable<string> directories, ReaderParameters parameters)
            {
                var extensions = name.IsWindowsRuntime ? new[] { ".winmd", ".dll" } : new[] { ".exe", ".dll" };
                foreach (var directory in directories)
                {
                    foreach (var extension in extensions)
                    {
                        var file = Path.Combine(directory, name.Name + extension);
                        if (!File.Exists(file))
                            continue;
                        try
                        {
                            return GetAssembly(file, parameters);
                        }
                        catch (BadImageFormatException)
                        {
                        }
                    }
                }

                return null;
            }

            private AssemblyDefinition GetAssembly(string file, ReaderParameters parameters)
            {
                if (parameters.AssemblyResolver == null)
                    parameters.AssemblyResolver = this;

                return ModuleDefinition.ReadModule(file, parameters).Assembly;
            }
        }
    }
}
